bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/util/DirectPathUtil.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.util;
+
+import com.google.api.core.InternalApi;
+import com.google.cloud.bigtable.config.BigtableOptions;
+import com.google.cloud.bigtable.config.CredentialOptions;
+import com.google.cloud.bigtable.config.CredentialOptions.CredentialType;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.io.CharStreams;
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+@InternalApi("For internal use only")
+public class DirectPathUtil {
+  private static final String GCE_PRODUCTION_NAME_PRIOR_2016 = "Google";
+  private static final String GCE_PRODUCTION_NAME_AFTER_2016 = "Google Compute Engine";
+
+  private static final Supplier<Boolean> isOnComputeEngine =
+      Suppliers.memoize(
+          new Supplier<Boolean>() {
+            @Override
+            public Boolean get() {
+              String osName = System.getProperty("os.name");
+              if ("Linux".equals(osName)) {
+                String cmd = "cat /sys/class/dmi/id/product_name";
```

Comment:
drive-by: 

why `cat` and then read the output?  That file (if it exists) `/sys/class/dmi/id/produc_name` can be opened like any other file.

Also: I think this won't work on GKE, the containers in that case run without the necessary permissions to read that file.

Reply:
These are excellent points. I was trying to keep consistent with all of the other java clients:
https://github.com/googleapis/gax-java/pull/1250

It seems like that file is accessible in gke:
https://github.com/googleapis/gax-java/pull/1250/files/cf494cef60f4d9ef265e5b4690ebe3cb49537893#r589861385

I'll update both libraries to simply read the file w/o exec

Reply:
Ooops, my recollection was wrong regarding GKE, it was Cloud Run and gVisor, but that may have changed again:

https://github.com/googleapis/google-cloud-cpp/pull/3959

For these purposes it might not matter because using Direct Path with Cloud Run might not be that useful or important.  If you are ever interested in supporting Windows, I wrote the code a long time ago for that:

https://github.com/grpc/grpc/blob/master/src/core/lib/security/credentials/alts/check_gcp_environment_windows.cc

Reply:
Thanks for the pointer. Though registry access from java is a tough pill to swallow.